### DISCUSSION TITLE
[Proposal] Terse Object Initializer Syntax

### DISCUSSION THREAD

**Discussion by leandromoh** [dotnet/csharplang#4961] [Language Ideas]
when creating anonymous object and tuples we can do something like

```c#
var foo = new { x.Name, y.Age };
```

which is just a suggar for the current

```c#
var foo = new { Name = x.Name, Age = y.Age };
```

Should be very interesting to have the same suggar for "normal" classes and structs and remove the current boilerplate, especially when they have lots of members. Example:

```c#
var p = new Person { x.Name, y.Age };
```

which is just a suggar for the current
```c#
var p = new Person { Name = x.Name, Age = y.Age };
```

obviously this suggar can only be applied where name and type of member match

**Comment by alrz** [comment]
>`var p = new Person { x.Name, y.Age };`

This is already valid for an enumerable type. e.g. `new List<object> { x.Name, y.Age }`. 

Due to this ambiguity, you have to spell out the member name in object initializers (unlike anonymous types).

**Reply by leandromoh** [reply]
Yeah, but this ambiguity only exists when the type is an enumrable right?  
Since most user defined types are not enumerable types, still looks worth to add the suggar even if strict for non-enumerable types.

**Reply by alrz** [reply]
In my opinion, the conceptual difference of the exact same syntax based on the details of a type definition doesn't seem like a good idea.

**Reply by leandromoh** [reply]
I understand, but looks to me that in fact this ambiguity would only happen in a very specific case: enumerable types that have settable members with the same name and type of the used values. That is, in `new List<object> { x.Name, y.Age }` since compiler knows that `List<>` does not have no public settable members Name (string) and Age (int) compiler could discard member initialization. 

And for the very specific case that the ambiguity exists, we could raise error CS0747 that alredy happen when you try to mix collection and member initialize in same initialization [both works, but not together](https://sharplab.io/#v2:EYLgtghglgdgNAFxAJwK7wCYgNQB8ACATAAwCwAUAASUVECMtxl+dALANyPN0B0AwgHsANkICmAYwRQBMAM6dyFaspXKAbhGSVxqYKMoBeSjFEB3SgAUBsBLKWqHj1QG97T9+4D0nymCgAPX1EwPS0IGAxtYTFJaRhKWCgpCCEoAC9RZDcPHJVvSmRoWX0+AGViAHZWCuzc3Nq6j3znSgBaOjg2jq7KAF9OhsanfIAxAQFDSgBmQaHqXoV6qhyAMwFkUQhxAAtKAAorGymAEUoABwT4nT0ASlmnFgBOPZZiHjHkSAQ9gCJnYl6lGcdEBzkIvR+nTOPAAGlCeABNeEALRuNwUtDoz2uone43RFGy+CmlFkCDQkks1hgCBO2VcywcxIKmwwMiEAE9KGzdGJKDDFo5mRsIGyYJzuQJefoEYKmSSRWKJTzgHzkRjGapmYcaSc9iq+f5Ogb9BzjVLVfo0ndNSoGbkYZN/HL3AjJhyXU5kZM0p7lL1sgHFJrmUQqTZZJQQNwpgAeHW044APnpszOyCgGgQ+kV7K5ABkoGT49TE0nKAB9M6l+SzZk2ShjAQah4khOyPY2pz2nJVmuTEzmQvFhMnJOdv3zQm25Sh1iUACCGAw+othvN0soZslm+tsx7Hj7EZ4S5Xg/DuuOeyNW861oJM6nwdbzHnp4OpZO5y7jgP7iPNKyCey57GcD5OEGdYkgAkgAojAqBgJkEAIOsJZHMmlAAOKiAg8GIchqHIJ2+73LOFSVtWx44XhCFIYURETrMkGPpQcF0YR6wxjwNH4fRKHrCRrF/g8FG8RxDGCeBjhBiovRAA). So user can simply not to use the suggar in this very very specific case that collections have public settable members with the same name and type of the used values.

**Reply by 333fred** [reply]
> enumerable types that have settable members with the same name and type of the used values

Which would mean that adding IEnumerable to a type would become a source-breaking change. This would not be a good idea, imo.

**Reply by leandromoh** [reply]
aaah, you have a point.. better avoid it ... solved

**Comment by Takym** [comment]
I propose the syntax below to avoid from the ambiguity:
1. Use a `$` symbol.
    ```csharp
    var p = new Person { $ = x.Name, $ = y.Age };
    ```
2. Use a `.` symbol.
    ```csharp
    var p = new Person { . = x.Name, . = y.Age };
    ```
2. Use a `from` operator.
    ```csharp
    var p = new Person { Name from x, Age from y };
    ```

**Reply by tacosontitan** [reply]
I'd go with `from` out of the three here. If that was the case however, I'd expand on it:

    var p = new Person {
        from x { Name },
        from y { Age, DateOfBirth }
    }

Simply due to the fact that the whole point is to gain brevity. Having to use `from` for each property individually doesn't really save us from anything we currently have to do.
